presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -214,33 +219,40 @@ public ScheduleResult schedule()
         List<Lifespan> driverGroupsToStart = ImmutableList.of();
         while (schedulerIterator.hasNext()) {
-            SourceScheduler sourceScheduler = schedulerIterator.next();
+            synchronized (this) {
+                // if a source scheduler is closed while it is scheduling, we can get an error
+                // prevent that by checking if scheduling has been cancelled first.
+                if (cancelled) {
+                    break;
+                }
+                SourceScheduler sourceScheduler = schedulerIterator.next();
 
-            for (Lifespan lifespan : driverGroupsToStart) {
-                sourceScheduler.startLifespan(lifespan, partitionHandleFor(lifespan));
-            }
+                for (Lifespan lifespan : driverGroupsToStart) {
+                    sourceScheduler.startLifespan(lifespan, partitionHandleFor(lifespan));
+                }
 
-            ScheduleResult schedule = sourceScheduler.schedule();
-            if (schedule.getSplitsScheduled() > 0) {
-                stage.transitionToSchedulingSplits();
-            }
-            splitsScheduled += schedule.getSplitsScheduled();
-            if (schedule.getBlockedReason().isPresent()) {
-                blocked.add(schedule.getBlocked());
-                blockedReason = blockedReason.combineWith(schedule.getBlockedReason().get());
-            }
-            else {
-                verify(schedule.getBlocked().isDone(), "blockedReason not provided when scheduler is blocked");
-                allBlocked = false;
-            }
+                ScheduleResult schedule = sourceScheduler.schedule();
+                if (schedule.getSplitsScheduled() > 0) {
+                    stage.transitionToSchedulingSplits();
+                }
+                splitsScheduled += schedule.getSplitsScheduled();
+                if (schedule.getBlockedReason().isPresent()) {
+                    blocked.add(schedule.getBlocked());
+                    blockedReason = blockedReason.combineWith(schedule.getBlockedReason().get());
+                }
+                else {
+                    verify(schedule.getBlocked().isDone(), "blockedReason not provided when scheduler is blocked");
+                    allBlocked = false;
+                }
 
-            driverGroupsToStart = sourceScheduler.drainCompletelyScheduledLifespans();
+                driverGroupsToStart = sourceScheduler.drainCompletelyScheduledLifespans();
 
-            if (schedule.isFinished()) {
-                stage.schedulingComplete(sourceScheduler.getPlanNodeId());
-                schedulerIterator.remove();
-                sourceScheduler.close();
-                anySourceSchedulingFinished = true;
+                if (schedule.isFinished()) {
+                    stage.schedulingComplete(sourceScheduler.getPlanNodeId());
+                    sourceSchedulers.remove(sourceScheduler);
+                    sourceScheduler.close();
```

Comment:
I presume it's fine for there to be duplicate closes

Reply:
Meaning one thread could be in the synchronized `FixedSourcePartitionedScheduler#close` method, but because this remove call is not synchronized, `FixedSourcePartitionedScheduler#close` could be called after the synchronized block above but before this line.  So hopefully duplicate close is OK here.